department-of-veterans-affairs/caseflow
# CAVC Dashboard: Ability to select multiple Basis for Selections for a Decision Reason

Resolves https://vajira.max.gov/browse/APPEALS-17475

### Description
- Adds table cavc_reasons_to_bases and associated model cavc_reasons_to_basis
- Modifies ActiveRecord relations between dashboard models to link them properly and cascade deletion where needed
- Modifies cavc_dashboard_disposition_serializer and implements cavc_dispositions_to_reason_serializer for returning data from the backend
- Refactor Basis for selection components out of CavcDecisionReasons.jsx and into CavcSelectionBasis.jsx
- Modify `initialCheckboxes` in CavcDecisionReasons to allow for an array of selection_bases instead of a single basis_for_selection on both parent and child decision reasons
- Modify all instances of using `setCheckedReasons` to work with an array of selection_bases
  - handleBasisChange, handleOtherTextFieldChange, handleCheckboxChange
- Add 'Add basis' and 'Remove basis' buttons to add/remove selection bases to a decision reason
- Add `handleAddBasis` and  `handleRemoveBasis` to CavcDecisionReasons for adding a selection basis to a decision reason
- Modify actions, reducer to support array of selection bases
- Style changes
- Test updates

### Acceptance Criteria

### Testing Plan
File numbers for dashboards with pre-populated selection bases: 410_000_006 through 410_000_010, 410_000_021

1. Go to any CAVC dashboard as an OAI user
2. Set any issue's disposition to Vacated and Remanded or Reversed
3. Check the box for one of the decision reasons that triggers the selection basis dropdown
4. Click 'Add basis'
5. Select any desired value for the selection basis
6. Click 'Add basis' for the same decision reason
7. Select any desired value for the selection basis
8. Repeat this as many times as desired
9. Click 'Save changes'
10. Go back into the dashboard
11. Verify that all selection bases that were added above are present on their associated decision reason
12. Remove any number of those selection bases
13. Click 'Save changes'
14. Go back into the dashboard
15. Verify that all removed selection bases that were removed above are no longer present
16. Go to any dashboard as an OAI user
17. Set any issue's disposition to Vacated and Remanded or Reversed
18. Check the box for one of the decision reasons that triggers the selection basis dropdown
19. Click 'Add basis'
20. Select any desired value for the selection basis
21. Click 'Save changes'
22. Go to the above dashboard as an OCC user
23. Verify that all information is read only, and that the 'Add basis' and 'Remove basis' buttons do not appear

### User Facing Changes

 BEFORE
<img width="620" alt="image" src="https://user-images.githubusercontent.com/109101548/226702612-44cbdd61-fc6b-4891-9757-510751ab9b95.png">
<img width="617" alt="image" src="https://user-images.githubusercontent.com/109101548/226702683-84f9e850-ca27-419c-b428-9a59de928452.png">
AFTER
<img width="647" alt="image" src="https://user-images.githubusercontent.com/109101548/226702247-21c38827-e71f-436b-a02b-dee13c6e0698.png">
<img width="657" alt="image" src="https://user-images.githubusercontent.com/109101548/226702397-79517b24-12d5-4475-8360-201e274738a8.png">

### Code Documentation Updates

### Database Changes
*Only for Schema Changes*
